File: flut/.flutter/lib/flutter/animation/animation_controller.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flut/error.dart';

class FlutAnimationBehavior extends FlutEnumObject<AnimationBehavior> {
  const FlutAnimationBehavior()
    : super('AnimationBehavior', const {
        'normal': AnimationBehavior.normal,
        'preserve': AnimationBehavior.preserve,
      });
}

class FlutAnimationController with FlutRealtimeObject<AnimationController> {
  FlutAnimationController.createFromData({
    required FlutRuntime runtime,
    required Map<String, dynamic> data,
    required AnimationController target,
  }) {
    initRealtimeFromData(runtime: runtime, data: data, target: target);
  }

  FlutAnimationController.createFromObject({
    required FlutRuntime runtime,
    required int oid,
    required AnimationController target,
  }) {
    initRealtimeFromObject(
      runtime: runtime,
      oid: oid,
      type: 'AnimationController',
      target: target,
    );
  }

  static FlutRealtimeObject flutCreate(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    final vsync = runtime.resolveArg(data['vsync']);
    if (vsync is! TickerProvider) {
      throw FlutRuntimeException('vsync object is not a TickerProvider');
    }
    final controller = AnimationController(
      value: runtime.unpackOptionalField<double>(data, 'value'),
      duration: runtime.unpackOptionalField<Duration>(data, 'duration'),
      reverseDuration: runtime.unpackOptionalField<Duration>(
        data,
        'reverseDuration',
      ),
      debugLabel: runtime.unpackOptionalField<String>(data, 'debugLabel'),
      lowerBound: runtime.unpackRequiredField<double>(data, 'lowerBound'),
      upperBound: runtime.unpackRequiredField<double>(data, 'upperBound'),
      animationBehavior: runtime.unpackRequiredField<AnimationBehavior>(
        data,
        'animationBehavior',
      ),
      vsync: vsync,
    );
    return FlutAnimationController.createFromData(
      runtime: runtime,
      data: data,
      target: controller,
    );
  }

  @override
  dynamic getRawProperty(String property) {
    switch (property) {
      case 'value':
        return flutTarget.value;
      case 'status':
        return flutTarget.status;
      case 'isAnimating':
        return flutTarget.isAnimating;
      case 'duration':
        return flutTarget.duration;
      case 'reverseDuration':
        return flutTarget.reverseDuration;
    }
    throw FlutUnknownPropertyException('AnimationController', property);
  }

  @override
  bool setProperty(String property, dynamic value) {
    switch (property) {
      case 'value':
        flutTarget.value = (value as num).toDouble();
        return true;
      case 'duration':
        flutTarget.duration = value as Duration?;
        return true;
      case 'reverseDuration':
        flutTarget.reverseDuration = value as Duration?;
        return true;
    }
    return false;
  }

  @override
  dynamic callMethod(
    String method,
    List<dynamic> args,
    Map<String, dynamic> kwargs,
  ) {
    switch (method) {
      case 'forward':
        final from = kwargs['from_'] as double?;
        flutTarget.forward(from: from);
        return null;
      case 'reverse':
        final from = kwargs['from_'] as double?;
        flutTarget.reverse(from: from);
        return null;
      case 'repeat':
        flutTarget.repeat(
          min: kwargs['min'] as double?,
          max: kwargs['max'] as double?,
          reverse: kwargs['reverse'] as bool? ?? false,
          period: kwargs['period'] as Duration?,
        );
        return null;
      case 'stop':
        flutTarget.stop();
        return null;
      case 'reset':
        flutTarget.reset();
        return null;
      case 'animateTo':
        flutTarget.animateTo(
          (args[0] as num).toDouble(),
          duration: kwargs['duration'] as Duration?,
          curve: kwargs['curve'] as Curve? ?? Curves.linear,
        );
        return null;
    }
    throw FlutUnknownMethodException(method);
  }
}

File: flut/flutter/animation/animation_controller.py
from typing import Optional

from flut._flut.engine import FlutEnumObject, _flut_pack_value
from flut.flutter.foundation.change_notifier import Listenable
from flut.flutter.animation.animation import Animation, AnimationStatus
from flut.flutter.animation.curves import Curve, Curves
from flut.dart.core import Duration


class AnimationBehavior(FlutEnumObject):
    normal: "AnimationBehavior"
    preserve: "AnimationBehavior"


class AnimationController(Animation):
    _flut_type = "AnimationController"

    def __init__(
        self,
        *,
        value: Optional[float] = None,
        duration: Optional[Duration] = None,
        reverseDuration: Optional[Duration] = None,
        debugLabel: Optional[str] = None,
        lowerBound: float = 0.0,
        upperBound: float = 1.0,
        animationBehavior: AnimationBehavior = AnimationBehavior.normal,
        vsync,
    ):
        props = {
            "lowerBound": lowerBound,
            "upperBound": upperBound,
            "vsync": vsync._flut_pack(),
            "animationBehavior": _flut_pack_value(animationBehavior),
        }
        if value is not None:
            props["value"] = value
        if duration is not None:
            props["duration"] = duration._flut_pack()
        if reverseDuration is not None:
            props["reverseDuration"] = reverseDuration._flut_pack()
        if debugLabel is not None:
            props["debugLabel"] = debugLabel
        self._flut_init_props = props
        Listenable.__init__(self)

    @property
    def value(self) -> float:
        return self._flut_get("value")

    @value.setter
    def value(self, new_value: float):
        self._flut_set("value", new_value)

    @property
    def status(self) -> AnimationStatus:
        return self._flut_get("status")

    @property
    def isAnimating(self) -> bool:
        return bool(self._flut_get("isAnimating"))

    @property
    def duration(self) -> Optional[Duration]:
        return self._flut_get("duration")

    @duration.setter
    def duration(self, value: Optional[Duration]):
        self._flut_set("duration", value._flut_pack() if value is not None else None)

    @property
    def reverseDuration(self) -> Optional[Duration]:
        return self._flut_get("reverseDuration")

    @reverseDuration.setter
    def reverseDuration(self, value: Optional[Duration]):
        self._flut_set(
            "reverseDuration", value._flut_pack() if value is not None else None
        )

    def forward(self, *, from_: Optional[float] = None):
        if from_ is not None:
            self._flut_call("forward", from_=from_)
        else:
            self._flut_call("forward")

    def reverse(self, *, from_: Optional[float] = None):
        if from_ is not None:
            self._flut_call("reverse", from_=from_)
        else:
            self._flut_call("reverse")

    def repeat(
        self,
        *,
        min: Optional[float] = None,
        max: Optional[float] = None,
        reverse: bool = False,
        period: Optional[Duration] = None,
    ):
        kwargs = {"reverse": reverse}
        if min is not None:
            kwargs["min"] = min
        if max is not None:
            kwargs["max"] = max
        if period is not None:
            kwargs["period"] = period._flut_pack()
        self._flut_call("repeat", **kwargs)

    def stop(self):
        self._flut_call("stop")

    def reset(self):
        self._flut_call("reset")

    def animateTo(
        self,
        target: float,
        *,
        duration: Optional[Duration] = None,
        curve: Curve = Curves.linear,
    ):
        kwargs = {"curve": curve._flut_pack()}
        if duration is not None:
            kwargs["duration"] = duration._flut_pack()
        self._flut_call("animateTo", target, **kwargs)
